### ISSUE TITLE
[BUG]: Crash on Parallel Loop over Simple Struct objects

### ISSUE THREAD

**Issue by ErenBalatkan** [modularml/mojo#1481]
### Bug description

Gives following error highlighted at top of the file on VSCode;
``
A crash happened in the mojo parser with the current version of this file. Please report this issue in https://github.com/modularml/mojo/issues along with all the relevant source codes with their current contents.mojo
``

And following error when I try to run the code;
```
/usr/bin/../lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/optional:438: _Tp &std::_Optional_base_impl<mlir::OpBuilder, std::_Optional_base<mlir::OpBuilder, true, true>>::_M_get() [_Tp = mlir::OpBuilder, _Dp = std::_Optional_base<mlir::OpBuilder, true, true>]: Assertion 'this->_M_is_engaged()' failed.
[13201:13201:20231213,182858.387742:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_cur_freq: No such file or directory (2)
[13201:13201:20231213,182858.387782:ERROR file_io_posix.cc:144] open /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq: No such file or directory (2)
Please submit a bug report to https://github.com/modularml/mojo/issues and include the crash backtrace along with all the relevant source codes.
Stack dump:
0.      Program arguments: /home/eren/.modular/pkg/packages.modular.com_mojo/bin/mojo /home/eren/hello.mojo
1.      Crash resolving decl body at loc("/home/eren/hello.mojo":18:1)
    >> fn main():
       ^.........
    >>     var Chars = DynamicVector[GameCharacter]()
    >>     for i in range(11):
    >>         Chars.append(GameCharacter(i))
    >> 
2.      Crash parsing statement at loc("/home/eren/hello.mojo":37:5)
    >>     parallelize[DealDamageParallel](ThreadCount)
           ^...........................................<
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  mojo      0x000056330c35ee17
1  mojo      0x000056330c35c9ee
2  mojo      0x000056330c35f4ef
3  libc.so.6 0x00007f1cd2ecc520
4  libc.so.6 0x00007f1cd2f209fc pthread_kill + 300
5  libc.so.6 0x00007f1cd2ecc476 raise + 22
6  libc.so.6 0x00007f1cd2eb27f3 abort + 211
7  mojo      0x000056330c803218
8  mojo      0x000056330c7fbeab
9  mojo      0x000056330c7f6ec9
10 mojo      0x000056330c8074e7
11 mojo      0x000056330c7fda1e
12 mojo      0x000056330c8109d2
13 mojo      0x000056330c80eb8c
14 mojo      0x000056330c7fd4b9
15 mojo      0x000056330c80bcbb
16 mojo      0x000056330c7fd789
17 mojo      0x000056330c836a1d
18 mojo      0x000056330c831148
19 mojo      0x000056330c830ff6
20 mojo      0x000056330c7cce8c
21 mojo      0x000056330c7e83b3
22 mojo      0x000056330c7e8c26
23 mojo      0x000056330c78cae9
24 mojo      0x000056330c78ce5b
25 mojo      0x000056330c32801f
26 mojo      0x000056330c314cfb
27 mojo      0x000056330c326ecf
28 mojo      0x000056330c30e3b6
29 libc.so.6 0x00007f1cd2eb3d90
30 libc.so.6 0x00007f1cd2eb3e40 __libc_start_main + 128
31 mojo      0x000056330c30d97e
Aborted
```

### Steps to reproduce

I'm simply attaching the entire code, it works perfectly fine if I remove parallel parts. 

```
from utils.vector import DynamicVector
from algorithm import parallelize
import math

trait Damageable:
    fn TakeDamage(inout self, Damage: FloatLiteral):...

@value
struct GameCharacter(Damageable, Stringable, CollectionElement):
    var Health: FloatLiteral

    fn TakeDamage(inout self, Damage: FloatLiteral):
        self.Health -= Damage

    fn __str__(self) -> String:
        return self.Health

fn main():
    var Chars = DynamicVector[GameCharacter]()
    for i in range(11):
        Chars.append(GameCharacter(i))

    let ThreadCount = 2
    let ElemPerThread : Int = Chars.size // ThreadCount

    fn DealDamageParallel(ThreadIdx : Int):
        let Start = ElemPerThread * ThreadIdx
        var End = ElemPerThread * (ThreadIdx + 1)
        if ThreadIdx == ThreadCount - 1:
            End = Chars.size

        print(Start, End, ElemPerThread)

        for Offset in range(End - Start):
            print(Start + Offset)

    parallelize[DealDamageParallel](ThreadCount)
```

### System information

```shell
- What OS did you do install Mojo on ?
Ubuntu 22.04 WSL
- Provide version information for Mojo by pasting the output of `mojo -v`
mojo 0.6.0 (d55c0025)
- Provide Modular CLI version by pasting the output of `modular -v`
modular 0.2.2 (95d42445)
```

**Comment by HJLebbink** [comment]
You need to annotate with @parameter

```python
@parameter
fn DealDamageParallel(ThreadIdx : Int):
```

And yes, a more forgiving error message would have helped.

**Comment by laszlokindrat** [comment]
@zyx-billy seems like the parser is crashing where an error message should be raised. Could you please take a look?

**Comment by duckki** [comment]
It doesn't crash with mojo 0.6.1. It seems fixed in 0.6.1.
